include/swift/IDE/CodeCompletion.h

```diff
@@ -981,6 +966,22 @@ class CodeCompletionResult {
         DiagnosticMessage(DiagnosticMessage), NumBytesToErase(NumBytesToErase),
         TypeDistance(TypeDistance) {}
 
+  /// Enrich a \c ContextFreeCodeCompletionResult with the following contextual
+  /// information.
+  /// This computes the type relation between the completion item and its
+  /// expected type context.
+  /// The \c ContextFree result must outlive this result. Typically, this is
+  /// done by allocating the two in the same sink or adopting the context free
+  /// sink in the sink that allocates this result.
+  CodeCompletionResult(const ContextFreeCodeCompletionResult &ContextFree,
```

Comment:
Is there a strong reason we need this constructor? Why can't callers calculate `TypeDistance` and call the existing constructor?

Reply:
There’s no strong reason but I wanted to have a single entry point that computes the type relation. Previously, we had a few places that tried doing something clever when commuting the type relation like knowing that the type relation is identical when adding a `#selector` completion if we are in a selector context. That complicated caching (because we didn’t know the actual result type) and in a few cases wasn’t actually correct because it didn’t take optional conversions into account.

I actually fell for exactly the same mistake because we had the member wise initializer. So what I did was make the member wise initializer private and only offer the constructor which computes the type relation, publicly.